File: rb101/rb101-rb109-small_problems/medium2/triangle_sides.rb
def triangle(side1, side2, side3)
  sides = [side1, side2, side3].sort
  short_sum = sides[0] + sides [1]
  longest_side = sides[2]

  # Note the use of the ',' in the 'case' statement. It's the equivalent
  # of using || in an 'if' statement.
  case
  when short_sum < longest_side, sides.include?(0) then :invalid
  when sides.uniq.size == 1 then :equilateral
  when sides.uniq.size == 2 then :isosceles
  else
    :scalene
  end
end

p triangle(3, 3, 3) == :equilateral
p triangle(3, 3, 1.5) == :isosceles
p triangle(3, 4, 5) == :scalene
p triangle(0, 3, 3) == :invalid
p triangle(3, 1, 1) == :invalid

# ALTERNATIVE SOLUTION:

# def triangle(side1, side2, side3)
#   sides = [side1, side2, side3]
#   largest_side = sides.max

#   case
#   when 2 * largest_side > sides.reduce(:+), sides.include?(0)
#     :invalid
#   when side1 == side2 && side2 == side3
#     :equilateral
#   when side1 == side2 || side1 == side3 || side2 == side3
#     :isosceles
#   else
#     :scalene
#   end
# end

File: js210/js210-small_problems/medium2/triangleSides.js
"use strict";

function triangle(side1, side2, side3) {
  if (!isValidTriangle(side1, side2, side3)) return 'invalid';
  if (side1 === side2 && side2 === side3) return 'equilateral';
  if (side1 !== side2 && side2 !== side3) return 'scalene';
  return 'isosceles';
}

function isValidTriangle(side1, side2, side3) {
  let sortedSides = [side1, side2, side3].sort((a, b) => a - b);
  if (sortedSides.some(length => length <= 0)) return false;
  return sortedSides[0] + sortedSides[1] > sortedSides[2];
}

console.log(triangle(3, 3, 3));        // "equilateral"
console.log(triangle(3, 3, 1.5));      // "isosceles"
console.log(triangle(3, 4, 5));        // "scalene"
console.log(triangle(0, 3, 3));        // "invalid"
console.log(triangle(3, 1, 1));        // "invalid"


// ALTERNATIVE SOLUTION:

// function triangle(side1, side2, side3) {
//   const perimeter = side1 + side2 + side3;
//   const longest = Math.max(side1, side2, side3);
//   const shortest = Math.min(side1, side2, side3);
//   const middle = perimeter - longest - shortest;

//   if (isValid(shortest, middle, longest)) {
//     return getTriangleType(side1, side2, side3);
//   } else {
//     return 'invalid';
//   }
// }

// function isValid(shortest, middle, longest) {
//   return shortest > 0 && (shortest + middle > longest);
// }

// function getTriangleType(side1, side2, side3) {
//   if (side1 === side2 && side2 === side3) {
//     return 'equilateral';
//   } else if (side1 === side2 || side1 === side3 || side2 === side3) {
//     return 'isosceles';
//   } else {
//     return 'scalene';
//   }
// }
